coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java

```diff
@@ -829,11 +829,11 @@ private void flushCurrentBatch() {
         }
 
         /**
-         * Flushes the current batch if it is transactional or if it has passed the append linger time.
+         * Flushes the current batch if it is transactional, if it has passed the append linger time, or if it is full.
          */
         private void maybeFlushCurrentBatch(long currentTimeMs) {
             if (currentBatch != null) {
-                if (currentBatch.builder.isTransactional() || (currentBatch.appendTimeMs - currentTimeMs) >= appendLingerMs) {
+                if (currentBatch.builder.isTransactional() || (currentBatch.appendTimeMs - currentTimeMs) >= appendLingerMs || !currentBatch.builder.hasRoomFor(0)) {
```

Comment:
@majialoong and I were discussing the condition `(currentBatch.appendTimeMs - currentTimeMs) >= appendLingerMs`. The correct version seems to be `(currentTimeMs - currentBatch.appendTimeMs) >= appendLingerMs`. Or we can remove it, since a `lingerTimeoutTask` already exists

WDYT?